pkg/apiserver/metrics/deployments.go

```diff
@@ -0,0 +1,168 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"k8s.io/kubernetes/pkg/api/rest"
+	"k8s.io/kubernetes/pkg/apis/extensions"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"golang.org/x/net/context"
+)
+
+var (
+	variableLabels = []string{"name", "namespace"}
+
+	deploymentStatusObservedGenerationDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_observed_generation",
+		"The observed generation of this deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_replicas",
+		"The number of replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUpdatedReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_updated_replicas",
+		"The number of updated replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusAvailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_available_replicas",
+		"The number of available replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+
+	deploymentStatusUnavailableReplicasDesc = prometheus.NewDesc(
+		"kubernetes_deployment_status_unavailable_replicas",
+		"The number of unavailable replicas for the specific deployment.",
+		variableLabels,
+		nil,
+	)
+)
+
+// DeploymentCollector collects various metrics of deployment objects.
+type DeploymentCollector struct {
+	Storage rest.StandardStorage
+}
+
+// Returns a new instance of DeploymentCollector.
+func NewDeploymentsCollector(s rest.StandardStorage) *DeploymentCollector {
+	return &DeploymentCollector{
+		Storage: s,
+	}
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- deploymentStatusObservedGenerationDesc
+	ch <- deploymentStatusReplicasDesc
+	ch <- deploymentStatusUpdatedReplicasDesc
+	ch <- deploymentStatusAvailableReplicasDesc
+	ch <- deploymentStatusUnavailableReplicasDesc
+}
+
+// Describe implements the prometheus.Collector interface.
+func (c *DeploymentCollector) Collect(ch chan<- prometheus.Metric) {
+	var err error
+	deployments, err := c.Storage.List(context.Background(), nil)
+	if err != nil {
+		glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+	}
+
+	for _, d := range deployments.(*extensions.DeploymentList).Items {
+		err = addMetrics(ch, d)
+		if err != nil {
+			glog.Errorf("Error while collecting deployment metrics: %s", err.Error())
+		}
+	}
+}
+
+// addMetrics adds Metrics for a single deployment to the specified channel
+func addMetrics(ch chan<- prometheus.Metric, d extensions.Deployment) error {
```

Comment:
The controller approach sounds very good, I was expecting the "source of truth" to be etcd here hence the implementation in the apiserver. Nonetheless the "size" problem would still be present.

I've heard multiple times that things that are missing in terms of metrics within kubernetes is what people find out via kubectl (pod states, node states, deployment states). It seems to make sense to go that granular and then throw away dimensions as needed. There have even been a few probjects to expose these objects directly as prometheus metrics as an add-on exporter. However, that approach seems out of place as kubernetes is already instrumented with prometheus metrics, it just seems we need to find the correct spot for it.